language/move-prover/docgen/src/docgen.rs

```diff
@@ -0,0 +1,930 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#[allow(unused_imports)]
+use log::{info, warn};
+
+use itertools::Itertools;
+use num::BigUint;
+use regex::Regex;
+use serde::Serialize;
+use spec_lang::{
+    ast::{ModuleName, SpecBlockInfo, SpecBlockTarget},
+    code_writer::CodeWriter,
+    emit, emitln,
+    env::{FunctionEnv, GlobalEnv, ModuleEnv, Parameter, StructEnv, TypeConstraint, TypeParameter},
+    symbol::Symbol,
+    ty::TypeDisplayContext,
+};
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, VecDeque},
+    rc::Rc,
+};
+
+const KEYWORDS: &[&str] = &[
+    "abort",
+    "acquires",
+    "as",
+    "break",
+    "continue",
+    "copy",
+    "copyable",
+    "",
+    "else",
+    "false",
+    "fun",
+    "if",
+    "invariant",
+    "let",
+    "loop",
+    "module",
+    "move",
+    "native",
+    "public",
+    "resource",
+    "return",
+    "spec",
+    "struct",
+    "true",
+    "use",
+    "while",
+];
+
+const WEAK_KEYWORDS: &[&str] = &[
+    "apply",
+    "assert",
+    "assume",
+    "decreases",
+    "aborts_if",
+    "ensures",
+    "requires",
+    "pack",
+    "unpack",
+    "update",
+    "to",
+    "schema",
+    "include",
+    "old",
+    "except",
+    "global",
+    "mut",
+];
+
+/// Options passed into the documentation generator.
+#[derive(Debug, Clone, Serialize)]
+pub struct DocgenOptions {
+    /// The level where we start sectioning. Often markdown sections are rendered with
+    /// unnecessary large section fonts, setting this value high reduces the size.
+    pub section_level_start: usize,
+    /// Whether to include private functions in the generated docs.
+    pub include_private_fun: bool,
+    /// Whether to include specifications in the generated docs.
+    pub include_specs: bool,
+    /// Whether to put specifications in the same section as a declaration or put them all
+    /// into an independent section.
+    pub specs_inlined: bool,
+    /// Whether to include Move implementations.
+    pub include_impl: bool,
+    /// Max depth of entries in table-of-contents
+    pub toc_depth: usize,
```

Comment:
What happens if the depth is exceeded?

Reply:
Comment was misleading, corrected. This is the depth to which TOC entries are displayed, not which depth is allowed in general.